File: javascript/partition/partition.js

var randMin = 10;
var randMax = 100;
var pivot = (randMin + randMax) / 2;
var aryLen = 10;
var ary = new Array(aryLen)
  .fill(true)
  .map(e => (randMin + parseInt(Math.random() * (randMax - randMin))));
console.log(ary);

function partition(ary, pivot) {
  var left = -1;
  var right = ary.length;

  while (left < right) {
    while (left < right && ary[++left] < pivot);
    while (right > left && ary[--right] > pivot);

    var temp = ary[left];
    ary[left] = ary[right];
    ary[right] = temp;
  }

  return left;
}

console.log('pivot on', pivot);
var partAt = partition(ary, pivot);
console.log('part at', partAt);
console.log(ary);

File: python/partition/partition.py
import random

def partition(ary, pivot):
  left = -1
  right = len(ary)

  while left < right:
    while left < right:
      left = left + 1
      if ary[left] > pivot:
        break
    while right > left:
      right = right - 1
      if ary[right] < pivot:
        break

    temp = ary[left]
    ary[left] = ary[right]
    ary[right] = temp

  return left

randMin = 10
randMax = 100
pivot = (randMin + randMax) / 2
aryLen = 10
ary = random.sample(xrange(randMin, randMax + 1), aryLen)
print ary
print 'pivot on', pivot
partAt = partition(ary, pivot)
print 'part at', partAt
print ary